circuit_knitting/cutting/cut_finding/cco_utils.py

```diff
@@ -138,23 +138,20 @@ def greedy_best_first_search(
         Callable, search_space_funcs.next_state_func
     )
 
-    if search_space_funcs.goal_state_func(state, *args):
-        return state
-
-    best = min(
-        [
-            (search_space_funcs.cost_func(next_state, *args), k, next_state)
-            for k, next_state in enumerate(
-                search_space_funcs.next_state_func(state, *args)
-            )
-        ],
-        default=(None, None, None),
-    )
-
-    if best[-1] is None:  # pragma: no cover
-        # This covers a rare edge case.
-        # We have so far found no circuit that triggers it.
-        # Excluding from test coverage for now.
-        return None
-
-    return greedy_best_first_search(best[-1], search_space_funcs, *args)
+    while not search_space_funcs.goal_state_func(state, *args):
+        best = min(
+            [
+                (search_space_funcs.cost_func(next_state, *args), k, next_state)
+                for k, next_state in enumerate(
+                    search_space_funcs.next_state_func(state, *args)
+                )
+            ],
+            default=(None, None, None),
+        )
+        if best[-1] is None:  # pragma: no cover
+            # This covers a rare edge case.
+            # We have so far found no circuit that triggers it.
+            # Excluding from test coverage for now.
+            return None
+        state = best[-1]
```

Comment:
I prefer using new varnames (e.g. `state_out`) to modifying input varnames, unless I intend on modifying the input variable in-place in memory, in which case I would document it in the docstring. You simply reassign `state` to a new data structure in memory, which is fine, but I do try to avoid messing with the input variables altogether as a safety measure against accidentally modifying an input data structure.

What you have here is fine and safe, so it's up to you what you prefer.